### PR TITLE
Vectorized Numpy to Torch based Functions for SpecAugment

### LINKED ISSUE #10459: [Wav2Vec2] Improve SpecAugment function by converting numpy based function to pytorch based function
# 🚀 Feature request

As can be seen here: https://github.com/huggingface/transformers/blob/11655fafdd42eb56ad94e09ecd84d4dc2d1041ae/src/transformers/models/wav2vec2/modeling_wav2vec2.py#L47,

the function `_compute_mask_indices` (responsible for spec augment) of Wav2Vec2 is written in numpy which means that the function is not GPU compatible. 
The function could simply be rewritten in PyTorch, which should make training on GPU faster.

This "Good First Issue" is about converting `_compute_mask_indices` to PyTorch while keeping the same functionality. 

## Your contribution

I'm happy to guide the contributor through the PR!

### PR DESCRIPTION
**Pull request by 01-vyom**
# What does this PR do?

Fixes #10459 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@patrickvonplaten

### CODE REVIEW

**Comment by 01-vyom** [comment]
Following are the average test results for the `_compute_mask_indices` function when ran 100 times. Each X.1 subtest case is calculated with `attention_mask = None` and each X.2 subtest case is calculated with `attention_mask` calculated with the following code:
```
attention_mask = torch.ones((batch_size, sequence_length), device=torch_device, dtype=torch.long)
attention_mask[:, -sequence_length // 2 :] = 0
```
1) Test - 100 times
batch_size = 4
sequence_length = 60
mask_prob = 0.5
mask_length = 1
Test **1.1** - 
Result - seconds 
New Code GPU: 0.002872414588928223
New Code CPU: 0.0006633639335632324
Old Code: 0.0003826594352722168
Test **1.2** - 
Result - seconds
New Code GPU: 0.002973439693450928
New Code CPU: 0.0006422805786132813
Old Code: 0.0004153728485107422

2) Test - 100 times
batch_size = 100
sequence_length = 60
mask_prob = 0.5
mask_length = 1
Test **2.1** - 
Result - seconds 
New Code GPU: 0.0663988971710205
New Code CPU: 0.014422652721405029
Old Code: 0.008053600788116455
Test **2.2** - 
Result - seconds
New Code GPU: 0.06568058252334595
New Code CPU: 0.01404146671295166
Old Code: 0.008796172142028809

3) Test - 100 times
batch_size = 1000
sequence_length = 60
mask_prob = 0.5
mask_length = 1
Test **3.1** - 
Result - seconds 
New Code GPU: 0.6623778533935547
New Code CPU: 0.14311392545700075
Old Code: 0.08917582988739013
Test **3.2** - 
Result - seconds
New Code GPU: 0.6566315603256225
New Code CPU: 0.13569485664367675
Old Code: 0.08646429538726806

4) Test - 100 times
batch_size = 4
sequence_length = 1000
mask_prob = 0.5
mask_length = 1
Test **4.1** - 
Result - seconds 
New Code GPU: 0.0031879472732543944
New Code CPU: 0.0013749027252197266
Old Code: 0.00248842716217041
Test **4.2** - 
Result - seconds
New Code GPU: 0.0031322765350341795
New Code CPU: 0.0010571050643920898
Old Code: 0.0015622496604919434

5) Test - 100 times
batch_size = 4
sequence_length = 60
mask_prob = 0.5
mask_length = 4
Test **5.1** - 
Result - seconds 
New Code GPU: 0.003424525260925293
New Code CPU: 0.0008220672607421875
Old Code: 0.0003489851951599121
Test **5.2** - 
Result - seconds
New Code GPU: 0.0034962940216064454
New Code CPU: 0.0007469034194946289
Old Code: 0.0003824186325073242

6) Test - 100 times
batch_size = 4
sequence_length = 1000
mask_prob = 0.5
mask_length = 4
Test **6.1** - 
Result - seconds 
New Code GPU: 0.003502027988433838
New Code CPU: 0.0014672994613647461
Old Code: 0.0017711663246154786
Test **6.2** - 
Result - seconds
New Code GPU: 0.0034971165657043455
New Code CPU: 0.0011277437210083009
Old Code: 0.0011361241340637207

7) Test - 100 times
batch_size = 128
sequence_length = 1000
mask_prob = 0.5
mask_length = 4
Test **7.1** - 
Result - seconds 
New Code GPU: 0.10527128219604492
New Code CPU: 0.04762232780456543
Old Code: 0.052808206081390384
Test **7.2** - 
Result - seconds
New Code GPU: 0.1032623028755188
New Code CPU: 0.03513101100921631
Old Code: 0.03523270606994629

**Comment by patrickvonplaten** [comment]
Hey @01-vyom,

In general this looks good to me - I use your code to do a test fine-tuning run to ensure everything is fine. Think then we can merge soon :-)

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 1]
```diff
@@ -68,49 +69,41 @@ def _compute_mask_indices(
     <https://github.com/pytorch/fairseq/blob/e0788f7007a8473a76db573985031f3c94201e79/fairseq/data/data_utils.py#L376>`__.
     """
     bsz, all_sz = shape
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:71]: I know the names weren't good previously, but it would be very nice to give better names now. `bsz` -> `batch_size` and `all_sz` -> maybe `mask_length`?

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 2]
```diff
@@ -68,49 +69,41 @@ def _compute_mask_indices(
...
-        # add a random number for probabilistic rounding
-        mask_prob * all_sz / float(mask_length)
-        + np.random.rand()
-    )
-
-    all_num_mask = max(min_masks, all_num_mask)
-
-    mask_idcs = []
+    mask = torch.full((bsz, all_sz), False, device=device)
+    min_len_unique, mask_idcs = float("Inf"), []
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:73]:
```suggestion
    min_len_unique = float("inf")
    mask_indices = []
```

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 3]
```diff
@@ -68,49 +69,41 @@ def _compute_mask_indices(
...
     padding_mask = attention_mask.ne(1) if attention_mask is not None else None
+    if padding_mask is not None:
+        sz = all_sz - torch.sum(padding_mask, 1)
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:76]: what is `sz` ? -> maybe `seq_length_for_masking`

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 4]
```diff
@@ -68,49 +69,41 @@ def _compute_mask_indices(
...
+        num_mask = (((sz * mask_prob) / mask_length) + torch.rand(bsz, device=device)).int()
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:77]: Think `num_masks` is better to elaborate that it's a tensor

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 5]
```diff
@@ -68,49 +69,41 @@ def _compute_mask_indices(
...
+    else:
+        sz = torch.full([bsz], all_sz, device=device)
+        all_num_mask = ((all_sz * mask_prob) / mask_length) + torch.rand(1, device=device)[0]
+        num_mask = torch.full([bsz], all_num_mask, device=device, dtype=torch.int32)
+
+    num_mask[num_mask < min_masks] = min_masks
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:83]: maybe add a comment `# make sure masks is at least of length `min_masks`

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 6]
```diff
@@ -68,49 +69,41 @@ def _compute_mask_indices(
...
-            lengths[0] = min(mask_length, sz - 1)
-
-        min_len = min(lengths)
-        if sz - min_len <= num_mask:
-            min_len = sz - num_mask - 1
-
-        mask_idc = np.random.choice(sz - min_len, num_mask, replace=False)
-        mask_idc = np.asarray([mask_idc[j] + offset for j in range(len(mask_idc)) for offset in range(lengths[j])])
-        mask_idcs.append(np.unique(mask_idc[mask_idc < sz]))
+        curr_num_mask = int(num_mask[i])
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:85]: here I think it's better to have `num_mask = int(num_masks[i])`

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py [hunk 7]
```diff
@@ -68,49 +69,41 @@ def _compute_mask_indices(
...
+        curr_sz = sz[i]
```
Comment (patrickvonplaten) [inline src/transformers/models/wav2vec2/modeling_wav2vec2.py:86]: better names here as well

**Comment by 01-vyom** [comment]
> Hey @01-vyom,
> 
> Great job - thanks a lot for tackling this issue!
> 
> One question regarding the benchmarking it looks like CPU is faster than GPU -> is this actually the case? This would not be very good as it means that performing this computation on GPU would slow training down. Also the statistics about the "old code" does it refer to the old code being run on CPU or GPU?

Old-Code being run on CPu as it was only using Numpy.

**Comment by patrickvonplaten** [comment]
> While running with smaller size inputs i.e. inputs with lesser batch size and mask length, this is the case. But, as we increase our mash length and batch size, this implementation is much efficient. Also, while using GPU, I think it is ineffcient because the tensors are being transfered back and forth from CPU and GPU. Old-code being run on CPU as it was only using numpy.

Let me run some benchmarks on my side to see :-) Ideally we don't have to run anything on CPU

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py (hunk 7 above)
Comment (01-vyom) [reply src/transformers/models/wav2vec2/modeling_wav2vec2.py:86]: I can do something like this:
`curr_seq_length_for_masking = seq_lengths_for_masking[i]`

**Comment by 01-vyom** [comment]
I have made all the changes. 
@patrickvonplaten any updates on the benchmarks?

**Comment by patrickvonplaten** [comment]
Hey @01-vyom,

I couldn't see a big speed-up really when running on GPU -> I think the goal should really be to completely get rid of the for loop. I implemented a simpler version here: https://github.com/huggingface/transformers/pull/11764

which I'm benchmarking now. Will let you know!

**Comment by patrickvonplaten** [comment]
Hey @01-vyom,

Sorry I merged the PR now since it was more or less ready. I'm sorry to not have used your PR here! I'm sure we'll find other cool possible contributions for Wav2Vec2 though soon - will ping you if I find something interesting :-)

**Comment by 01-vyom** [comment]
No problem. Learned a lot from this issue.
